test/unit/org/apache/cassandra/index/IndexStatusManagerTest.java

```diff
@@ -0,0 +1,393 @@
+
+            Builder minRequiredNodes(int required)
+            {
+                minRequiredNodes = required;
+                return this;
+            }
+
+            Builder replicationStrategy(AbstractReplicationStrategy strategy)
+            {
+                replicationStrategy = strategy;
+                return this;
+            }
+
+            Builder indexStatus(Map<InetAddressAndPort, Map<String, Index.Status>> status)
+            {
+                indexStatus = status;
+                return this;
+            }
+
+            Builder expected(EndpointsForRange endpoints)
+            {
+                expected = endpoints;
+                return this;
+            }
+
+            Testcase build()
+            {
+                return new Testcase(this);
+            }
+        }
+    }
+
+    // we have enough BUILD_SUCCEEDED endpoints: filterForQuery should prioritize them, even if the other endpoints'
+    // indexes are queryable but not built (like UNKNOWN)
+    @Test
+    public void filterForQuery_ShouldPrioritizeSuccessfulEndpoints()
+    {
+        runTest("exactly enough endpoints with BUILD_SUCCEEDED indexes",
+                new Testcase.Builder()
+                        .keyspace("ks1")
+                        .replicationStrategy(new NetworkTopologyStrategy("ks1", Map.of("DC", "5")))
+                        .indexStatus(Map.of(
+                                InetAddressAndPort.getByNameUnchecked("127.0.0.255"),
+                                Map.of(
+                                        "ks1.idx1", Index.Status.BUILD_SUCCEEDED,
+                                        "ks1.idx2", Index.Status.BUILD_SUCCEEDED,
+                                        "ks1.idx3", Index.Status.UNKNOWN
+                                ),
+                                InetAddressAndPort.getByNameUnchecked("127.0.0.254"),
+                                Map.of(
+                                        "ks1.idx1", Index.Status.BUILD_SUCCEEDED,
+                                        "ks1.idx2", Index.Status.BUILD_SUCCEEDED,
+                                        "ks1.idx3", Index.Status.BUILD_SUCCEEDED
+                                ),
+                                InetAddressAndPort.getByNameUnchecked("127.0.0.253"),
+                                Map.of(
+                                        "ks1.idx1", Index.Status.BUILD_SUCCEEDED,
+                                        "ks1.idx2", Index.Status.UNKNOWN,
+                                        "ks1.idx3", Index.Status.BUILD_SUCCEEDED
+                                )
+                        ))
+                        .minRequiredNodes(1)
+                        .expected(EndpointsForRange.of(
+                                full(InetAddressAndPort.getByNameUnchecked("127.0.0.254"))
+                        ))
+                        .build()
+        );
+
+        runTest("more than enough endpoints with BUILD_SUCCEEDED indexes",
```

Comment:
I'd make this its own top-level test, called perhaps `shouldPrioritizeMultipleSuccessfulEndpoints()`